### DISCUSSION TITLE
Do I need a middleware?

### DISCUSSION THREAD

**Discussion by markomitranic** [amannn/next-intl#781] [General]
Hey folks, hey @amannn thanks again for making this awesome package.

According to the docs, if we wish to skip on using middleware, we should use another package. I believe this isn't correct, but I'd love to hear why I'm wrong and learn from it. I understand that the middleware is more complex than I show here (https://github.com/amannn/next-intl/blob/main/packages/next-intl/src/middleware/middleware.tsx) but I think it is more complex due to all the "extra stuff" and I'd like to propose an alternative, simpler usage.

The way I see it, middleware performs 4 distinct actions (from the perspective of the user):
1. url locale matching
2. set locale in global state
3. header neagotiation
4. rewrites

A lot of projects don't use 3 and 4. In fact, even though it seems like a cool idea, it is generally a poor UX and SEO practice to do forcible redirects, and is used only when we legally must redirect users like that, and even then it is done based on geofencing. When it comes to asking users to change market or language, that is best done as a popup in a client component and has nothing to do with next-intl. My point - a lot of projects don't need 3 & 4 AND 3 & 4 make up for the majority of the complexity of middleware.

Which leaves us with a middleware that only performs numbers 1 & 2. There is nothing in them that I can see, that requires them to be in the middleware:
- **Number 1** it is already expressed as a pure functional interface `resolveLocale` and used by the middleware to chew up the request and config, and spew out a validated locale.
- **Number 2** (as you very well know) is tricky and can consist of 3 branches, depending on the resolution environment - `NextIntlClientProvider`, `unstable_setRequestLocale` and the request assignment that the middleware currently does.

So, where am I going with this? I'm not saying remove middleware, a lot of people want it. But what if people like me were to drop the middleware entirely (thus reducing costs) and in our layouts write something along the lines of:
1. `const {_, locale} = resolveLocale(request);`
2. `<NextIntlClientProvider locale={locale} />`

The first one is self explanatory, it would put the decisionmaking logic closer to us. The second one would have `next-intl` decide which environment we are in at the moment (i believe you do that by forking package paths at the moment), and choose if the locale goes into `unstable_setRequestLocale` or into a Request+Context.

Where am I wrong? :D

**Comment by amannn** [comment]
Hey, thanks for the kind words and for the curious question! :)

> According to the docs, if we wish to skip on using middleware, we should use another package.

Which section in the docs are you referencing here?

There's a section on the Next.js static export feature, mentioning usage without a middleware: [Usage without middleware (static export)](https://next-intl-docs.vercel.app/docs/routing/middleware#usage-without-middleware-static-export).

> But what if people like me were to drop the middleware entirely (thus reducing costs) and in our layouts write something along the lines of:
> 
> 1. `const {_, locale} = resolveLocale(request);`
> 2. `<NextIntlClientProvider locale={locale} />`

If you resolve the locale in a layout, it will require that your app only works with dynamic rendering—that also affects costs.


> the middleware is more complex

Absolutely is, I'd really like to refactor the code at some point, potentially breaking the responsibilities into clearer defined code paths! The middleware has [a pretty good test suite](https://github.com/amannn/next-intl/blob/main/packages/next-intl/test/middleware/middleware.test.tsx) that could help with such a refactor.

The middleware currently has these responsibilities:
1. Negotiating the locale
2. Setting a cookie to remember the previously matched locale
3. Handling redirects to add missing locale prefixes
4. Handling rewrites if a `localePrefix` other than `always` is used
5. Handling rewrites if localized pathnames are used
6. Attaching the alternate links header
7. (currently) passing the locale to Server Components (more in [the static rendering section](https://next-intl-docs.vercel.app/docs/getting-started/app-router#static-rendering), [hopefully gets removed at some point](https://github.com/amannn/next-intl/issues/663))

(hope I've got all)

Note that users can also [implement a custom middleware](https://github.com/amannn/next-intl/issues/609) if that suits the use case better.

**Reply by markomitranic** [reply]
Thanks for a thorough answer!

> If you resolve the locale in a layout, it will require that your app only works with dynamic rendering—that also affects costs.

Hmm, how so? Isn't the layout pre-rendered separately for each locale I define in `generateStaticParams`, therefore resulting in a bunch of different static layout renders? Same for the pages (I suspect I'd have to repeat this in pages according to your previous comments). I guess I'd have to try and do a mock test of this to properly understand what I'm missing.

> The middleware currently has these responsibilities:

I get the point, `1, 2, 3` wouldn't be relevant - but the others still would, and we simply can't do `4, 5` anywhere other than middleware (unlike `6, 7` which seem doable with effort). In the end, it is less dirty to manage them all in the middleware. :P

> Note that users can also https://github.com/amannn/next-intl/issues/609 if that suits the use case better.

Yeah, I've started diving deeper into that. Also found some examples in other issues and discussions. I think `next-intl` serves all the needs of my agency in general. But my unfortunate use case with the current client dictates that they have 20 market based domains, each with 1 or more languages, all with `localePrefix: "as-needed"` and localized pathnames. Obviously the latter part is doable with the current functionality, but the domains+locales setup doesn't seem doable at all and will need some middleware tinkering. Thanks for the tip!

**Reply by amannn** [reply]
> Hmm, how so? Isn't the layout pre-rendered separately for each locale I define in `generateStaticParams`, therefore resulting in a bunch of different static layout renders? Same for the pages (I suspect I'd have to repeat this in pages according to your previous comments). I guess I'd have to try and do a mock test of this to properly understand what I'm missing.

Well yes, it depends—I guess my comment was a bit ambiguous. `resolveLocale` chooses a locale based on [request-based information](https://next-intl-docs.vercel.app/docs/routing/middleware#locale-detection). As mentioned in the other comment, the pathname is _only_ available in the middleware, while e.g. reading from headers and cookies opts into dynamic rendering.

If you have pages within a `[locale]` folder and you simply read that param, then you can use `generateStaticParams`.

> But my unfortunate use case with the current client dictates that they have 20 market based domains, each with 1 or more languages, all with `localePrefix: "as-needed"` and localized pathnames.

Sounds pretty exciting, you should share a link to the project once you have something up and running! :)
